Merge ObjC categories into docs of the extended class

esad opened the pull request:
This is work-in-progress on #457.

The idea was to annotate `SourceDeclaration`s representing extensions (both swift and objc) with a link to the parent declaration that is being extended, and then use this as a deduplication key. This kind of unifies the existing approach for swift extensions with new behaviour needed for objc classes/categories.

Furthermore, we found it nice to have category name appear as a mark of the merged methods.

segiddins commented on lib/jazzy/sourcekitten.rb:
```diff
@@ -318,8 +318,8 @@ def self.deduplicate_declarations(declarations)
     def self.deduplication_key(decl)
-      if decl.type.swift_extensible? || decl.type.swift_extension?
-        [decl.usr]
+      if (decl.type.swift_extension? || decl.type.objc_category?) && !decl.extension_parent.nil?
```
maybe add a helper method for `extension_type?`

segiddins commented on lib/jazzy/sourcekitten.rb:
```diff
@@ -340,6 +340,14 @@ def self.merge_declarations(decls)
+        # unless they already have a mark
+        extensions.each do |ext|
+          ext.children.each do |c|
+            c.mark.name ||= (ext.extension_name || ext.name)
```
the parens here are unnecessary

pcantrell commented:
Thanks for this, Esad! Two families of questions as I digest the code:

First, what’s the rationale behind computing the `extension_parent`? It looks like you’re doing it just to replace the deduplication key of `[decl.usr]` with `deduplication_key(decl.extension_parent)`; is that correct? I’m guessing that’s because Obj-C extensions come back with USRs different from the type they extend…?

If so, I wonder if perhaps the regexp dance you did in your [first crack at this](https://github.com/PSPDFKit-labs/jazzy/commit/7f8d68ad0d098fe1418211f309d066009941030f) was a better plan. It takes a _lot_ of code in this PR to construct `extension_parent`. Writing an `extract_extended_type_usr` method might be preferable.

What do those Obj-C USRs look like? Can you provide some examples?

---

Second, I wonder if it might make sense to unify Swift & Obj-C a bit more, e.g. `extension?` as a shortcut for `swift_extension? || objc_category?`,  in order to simplify some of those tangled conditionals.

I vaguely remember @jpsim saying that he has a preference for keeping Swift things and Objective-C things somewhat separate, but if he’s amenable to some judicious cross-language generalizing, it could tidy things a bit.

If we _are_ going to keep things separate, then `extension_name` should be `category_name`, because extensions are anonymous in Swift.

Nice idea using the category name as a marker, BTW.

esad commented:
Hi @pcantrell, thanks for your feedback!

Yes, the purpose of going through all extensions to find and set the `extension_parent` is the deduplication - it struck me as more elegant to simply say "for this extension, use the dedup key of the extended type" than to manually make it match. The `extension_name` is later used when setting the mark for the merged methods.

The USRs itself in Objective-C look like `c:objc(cy)SomeClass@SomeCategory` for the category and `c:objc(cs)SomeClass` for the class itself.

I'll try to refactor this back to a more simple solution, let's see how it looks.

esad commented:
I've updated the PR with a simpler approach.

segiddins commented:
Before merging, this will need a rebase to get rid of the merge commit

segiddins commented on lib/jazzy/source_declaration.rb:
```diff
@@ -41,6 +41,13 @@ def fully_qualified_name
+    # of the extended objc class and the category name itself, i.e.
+    # ["NSString", "MyMethods"], nil otherwise.
+    def objc_category_name
+      type.objc_category? && name.split(/[\(\)]/) || nil
```
This could just be `name.split(/[\(\)]/) if type.objc_category?` is

segiddins commented on lib/jazzy/sourcekitten.rb:
```diff
@@ -320,14 +320,18 @@ def self.deduplicate_declarations(declarations)
     end
 
     # Merges all of the given types and extensions into a single document.
+    # rubocop:disable Metrics/MethodLength
```
this should either be re-enabled after the method, or this should go on the same line as the method declaration

segiddins commented:
This warrants adding categories to an integration spec, I think

esad commented:
I addresses the above PR remarks, but I'm blocked by integration specs not passing on the current master. Waiting to resolve this first before I update them.

pcantrell commented:
I like the way this ended up. Not too complicated; things makes sense. I'm dropping a few minor notes into the code, and after those I'd say it's good to merge.

+1 to Sam’s comment that this warrants an addition to the specs, perhaps just a small addition to `misc_jazzy_features` — though I think, given this project’s somewhat loose coupling between features and tests, that could be a separate PR if it proves to be a hassle.

pcantrell commented on lib/jazzy/source_declaration.rb:
```diff
@@ -41,6 +41,13 @@ def fully_qualified_name
+    # of the extended objc class and the category name itself, i.e.
+    # ["NSString", "MyMethods"], nil otherwise.
+    def objc_category_name
+      name.split(/[\(\)]/) if type.objc_category?
```
The name `objc_category_name` implies that this returns a string, not an array — and that seems like the right behavior.

esad replied:
Do you think I should split this method into two: one to return the category name and the other to return the extended class name? Or rename the above one to match the method signature?

pcantrell replied:
Now that I understand better how you’re using it, I'd suggest renaming it to make it clear that it returns two values.

pcantrell commented on lib/jazzy/sourcekitten.rb:
```diff
@@ -320,14 +320,18 @@ def self.deduplicate_declarations(declarations)
       end
     end
 
+    # rubocop:disable Metrics/MethodLength
```
This method might benefit from a refactoring pass in the future. Right to just disable this for now and get it out the door, though!

esad replied:
Yep, I was thinking of taking care of this now, but wanted to keep the scope of this PR minimal

pcantrell commented on lib/jazzy/sourcekitten.rb:
```diff
@@ -320,14 +320,18 @@ def self.deduplicate_declarations(declarations)
       if decl.type.swift_extensible? || decl.type.swift_extension?
         [decl.usr, decl.name]
+      elsif decl.type.objc_class? || decl.type.objc_category?
+        name, = decl.objc_category_name || decl.name
```
Extraneous comma here?

segiddins replied:
nope, the comma allows this to extract the first element of an array

esad replied:
No, it's multiple assignment :) Rubocop complained per http://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/TrailingUnderscoreVariable so I changed it to match the required style.

pcantrell replied:
Stupid Rubocop. But now I see why you have objc_category_name returning two things!

pcantrell commented on lib/jazzy/sourcekitten.rb:
```diff
@@ -320,14 +320,18 @@ def self.deduplicate_declarations(declarations)
         [decl.usr, decl.name]
+      elsif decl.type.objc_class? || decl.type.objc_category?
+        name, = decl.objc_category_name || decl.name
+        [name, 'objc_class_and_categories']
```
I like this solution — maybe a little bit of a hack, but much tighter than manually resolving the relationships!

I'd make the second array elem a symbol.

esad commented:
@segiddins What would be the best way forward to add integration tests for the above? I tried adding some objc categories to the MiscJazzyFeatures project, but the SourceKitten ignores those (even though I added them to umbrealla header). It seems that jazzy currently doesn't support mixed Swift/ObjC projects? Should I add  a new MiscJazzyObjcFeatures project to the integration specs?

segiddins commented:
Yeah, I think a new `MiscJazzyObjcFeatures` project would probably be the best bet

esad commented:
I've added integration specs for this feature. As pointed out in the child PR, I'm bit unsure how this workflow with 2 dependent separate repos should work. Can you merge the child PR first, so that I can point to it here and we can have CI pass on this PR?

esad commented:
After integration specs pass I'll address the remaining code style issues.

pcantrell commented:
The integration specs process is a bit of a mystery to me too at times, but I merged your PR and hopefully that unblocks this merge.

As you do the last little tidying work, please feel free to edit `.rubocop.yml` to make it so you can do `name, _ =` instead of `name, =`. Anybody else feel free to pipe up in disagreement, but I find the latter terrible.

Thanks for bearing with this whole process!

esad commented:
@pcantrell @segiddins I've addressed all issues we discussed above, added integration specs and rebased the PR. Let me know if you feel there's something else missing.

esad commented:
I just wanted to let you know that we've rebuilt [PSPDFKit docs](https://pspdfkit.com/api/ios/) with this patch applied and it worked fine, with categories getting merged nicely into the class docs.

pcantrell commented:
I'd love to get this merged! It's only blocked on the CI failing. It looks like the integration specs are building Moya with a slightly different version of AF that the expected output recorded.

I have to speed-prep a talk I'm doing tonight, but @esad (or anyone else), if you have a chance to update the expected output, I can approve the merges.

esad commented:
@pcantrell As soon as https://github.com/realm/jazzy-integration-specs/pull/11/commits in jazzy-integration-specs gets merged I will update this PR

esad commented:
@pcantrell Looks like the specs are passing again

pcantrell commented:
Merging! Hooray!

Thanks for all the hard work and the patience.

jpsim commented:
This is great! Thanks @esad for your hard work, and @pcantrell & @segiddins for the reviews! :+1:
